runtime/compiler/control/CompilationThread.cpp

```diff
@@ -2596,6 +2618,60 @@ void TR::CompilationInfo::resumeCompilationThread()
    releaseCompMonitor(vmThread);
    }
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+void TR::CompilationInfo::prepareForCheckpoint()
+   {
+   J9JavaVM   *vm       = _jitConfig->javaVM;
+   J9VMThread *vmThread = vm->internalVMFunctions->currentVMThread(vm);
+
+   {
+   OMR::CriticalSection suspendCompThreadsForCheckpoint(getCompilationMonitor());
+
+   /* Set the checkpoint in progress flag. */
+   setCheckpointInProgress();
+
+   /* Inform compilation threads to suspend themselves. The compilation
+    * threads will complete any in-progress compilations first.
+    */
+   bool purgeMethodQueue = false;
+   suspendCompilationThread(purgeMethodQueue);
+   }
+
+   /* Wait until all compilation threads are suspended. */
+   {
+   OMR::CriticalSection waitForThreadsToSuspend(getCRMonitor());
+
+   for (int32_t i = 0; i < getNumTotalCompilationThreads(); i++)
+      {
+      TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
+
+      if (!curCompThreadInfoPT->isDiagnosticThread())
+         {
+         while (curCompThreadInfoPT->getCompilationThreadState() != COMPTHREAD_SUSPENDED)
```

Comment:
For the most part I believe we can. The reason is because in `prepareForCheckpoint` we acquire the comp monitor to trigger the suspension of threads. The code also checks in places where new threads could be activated and prevents it form doing so, essentially all the places that check for `getDisableFurtherCompilation`.

That said, I think there is still a problem in case of  JIT Dump or during Shutdown, when we set the state to `COMPTHREAD_SIGNAL_TERMINATE`.

I'll have to look into those cases more.

Reply:
I spent a long time thinking about it, and I believe I've finally gotten all the cases covered. I've updated the code, the commit, and this issue's description with the details, but to your initial point about the compMonitor, the snapshotting thread now does acquire the compMonitor and plays nice with the other threads.